// @cartesia/cartesia-js/dist/api/resources/tts/types/EmotionDeprecated.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmotionDeprecated = void 0;
exports.EmotionDeprecated = {
    AngerLowest: "anger:lowest",
    AngerLow: "anger:low",
    Anger: "anger",
    AngerHigh: "anger:high",
    AngerHighest: "anger:highest",
    PositivityLowest: "positivity:lowest",
    PositivityLow: "positivity:low",
    Positivity: "positivity",
    PositivityHigh: "positivity:high",
    PositivityHighest: "positivity:highest",
    SurpriseLowest: "surprise:lowest",
    SurpriseLow: "surprise:low",
    Surprise: "surprise",
    SurpriseHigh: "surprise:high",
    SurpriseHighest: "surprise:highest",
    SadnessLowest: "sadness:lowest",
    SadnessLow: "sadness:low",
    Sadness: "sadness",
    SadnessHigh: "sadness:high",
    SadnessHighest: "sadness:highest",
    CuriosityLowest: "curiosity:lowest",
    CuriosityLow: "curiosity:low",
    Curiosity: "curiosity",
    CuriosityHigh: "curiosity:high",
    CuriosityHighest: "curiosity:highest",
};
