// @elevenlabs/elevenlabs-js/dist/serialization/types/MultiSourceConfigJson.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiSourceConfigJson = void 0;
const core = __importStar(require("../../core"));
const MergingStrategy_1 = require("./MergingStrategy");
const SourceConfigJson_1 = require("./SourceConfigJson");
exports.MultiSourceConfigJson = core.serialization.object({
    sourceNames: core.serialization.property("source_names", core.serialization.list(core.serialization.string()).optional()),
    sourceOverrides: core.serialization.property("source_overrides", core.serialization.list(SourceConfigJson_1.SourceConfigJson).optional()),
    mergingStrategy: core.serialization.property("merging_strategy", MergingStrategy_1.MergingStrategy.optional()),
    finalTopK: core.serialization.property("final_top_k", core.serialization.number().optional()),
    useDecomposition: core.serialization.property("use_decomposition", core.serialization.boolean().optional()),
    useReformulation: core.serialization.property("use_reformulation", core.serialization.boolean().optional()),
    synthesizeResponse: core.serialization.property("synthesize_response", core.serialization.boolean().optional()),
});
