// @elevenlabs/elevenlabs-js/dist/serialization/types/SourceRetrievalConfig.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceRetrievalConfig = void 0;
const core = __importStar(require("../../core"));
exports.SourceRetrievalConfig = core.serialization.object({
    name: core.serialization.string(),
    collectionName: core.serialization.property("collection_name", core.serialization.string()),
    dbName: core.serialization.property("db_name", core.serialization.string().optional()),
    enabled: core.serialization.boolean().optional(),
    kDense: core.serialization.property("k_dense", core.serialization.number().optional()),
    kKeyword: core.serialization.property("k_keyword", core.serialization.number().optional()),
    denseWeight: core.serialization.property("dense_weight", core.serialization.number().optional()),
    keywordWeight: core.serialization.property("keyword_weight", core.serialization.number().optional()),
    sourceWeight: core.serialization.property("source_weight", core.serialization.number().optional()),
    vectorIndexName: core.serialization.property("vector_index_name", core.serialization.string().optional()),
    embeddingField: core.serialization.property("embedding_field", core.serialization.string().optional()),
    contentField: core.serialization.property("content_field", core.serialization.string().optional()),
    filterField: core.serialization.property("filter_field", core.serialization.string().optional()),
    numCandidatesMultiplier: core.serialization.property("num_candidates_multiplier", core.serialization.number().optional()),
    resultFields: core.serialization.property("result_fields", core.serialization
        .record(core.serialization.string(), core.serialization.list(core.serialization.unknown()))
        .optional()),
});
