// @elevenlabs/elevenlabs-js/dist/serialization/types/UpdateToolCallUnitTestRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateToolCallUnitTestRequest = void 0;
const core = __importStar(require("../../core"));
const ConversationHistoryTranscriptCommonModelInput_1 = require("./ConversationHistoryTranscriptCommonModelInput");
const TestFromConversationMetadataInput_1 = require("./TestFromConversationMetadataInput");
const UnitTestToolCallEvaluationModelInput_1 = require("./UnitTestToolCallEvaluationModelInput");
const UpdateToolCallUnitTestRequestDynamicVariablesValue_1 = require("./UpdateToolCallUnitTestRequestDynamicVariablesValue");
exports.UpdateToolCallUnitTestRequest = core.serialization.object({
    fromConversationMetadata: core.serialization.property("from_conversation_metadata", TestFromConversationMetadataInput_1.TestFromConversationMetadataInput.optional()),
    dynamicVariables: core.serialization.property("dynamic_variables", core.serialization
        .record(core.serialization.string(), UpdateToolCallUnitTestRequestDynamicVariablesValue_1.UpdateToolCallUnitTestRequestDynamicVariablesValue.optional())
        .optional()),
    chatHistory: core.serialization.property("chat_history", core.serialization.list(ConversationHistoryTranscriptCommonModelInput_1.ConversationHistoryTranscriptCommonModelInput).optional()),
    toolCallParameters: core.serialization.property("tool_call_parameters", UnitTestToolCallEvaluationModelInput_1.UnitTestToolCallEvaluationModelInput.optional()),
    checkAnyToolMatches: core.serialization.property("check_any_tool_matches", core.serialization.boolean().optional()),
    name: core.serialization.string(),
});
