// @cartesia/cartesia-js/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CartesiaClient = void 0;
const Client_1 = require("./api/resources/apiStatus/client/Client");
const Client_2 = require("./api/resources/auth/client/Client");
const Client_3 = require("./api/resources/infill/client/Client");
const Client_4 = require("./api/resources/stt/client/Client");
const Client_5 = require("./api/resources/tts/client/Client");
const Client_6 = require("./api/resources/voiceChanger/client/Client");
const Client_7 = require("./api/resources/voices/client/Client");
class CartesiaClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    get apiStatus() {
        var _a;
        return ((_a = this._apiStatus) !== null && _a !== void 0 ? _a : (this._apiStatus = new Client_1.ApiStatus(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new Client_2.Auth(this._options)));
    }
    get infill() {
        var _a;
        return ((_a = this._infill) !== null && _a !== void 0 ? _a : (this._infill = new Client_3.Infill(this._options)));
    }
    get stt() {
        var _a;
        return ((_a = this._stt) !== null && _a !== void 0 ? _a : (this._stt = new Client_4.Stt(this._options)));
    }
    get tts() {
        var _a;
        return ((_a = this._tts) !== null && _a !== void 0 ? _a : (this._tts = new Client_5.Tts(this._options)));
    }
    get voiceChanger() {
        var _a;
        return ((_a = this._voiceChanger) !== null && _a !== void 0 ? _a : (this._voiceChanger = new Client_6.VoiceChanger(this._options)));
    }
    get voices() {
        var _a;
        return ((_a = this._voices) !== null && _a !== void 0 ? _a : (this._voices = new Client_7.Voices(this._options)));
    }
}
exports.CartesiaClient = CartesiaClient;
