// @cartesia/cartesia-js/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CartesiaTimeoutError = exports.CartesiaError = exports.CartesiaEnvironment = exports.StreamingSTTClient = exports.SttWebsocket = exports.WebPlayer = exports.CartesiaClient = exports.Cartesia = void 0;
exports.Cartesia = __importStar(require("./api"));
var Client_1 = require("./wrapper/Client");
Object.defineProperty(exports, "CartesiaClient", { enumerable: true, get: function () { return Client_1.CartesiaClient; } });
var WebPlayer_1 = require("./wrapper/WebPlayer");
Object.defineProperty(exports, "WebPlayer", { enumerable: true, get: function () { return WebPlayer_1.WebPlayer; } });
var SttWebsocket_1 = require("./wrapper/SttWebsocket");
Object.defineProperty(exports, "SttWebsocket", { enumerable: true, get: function () { return __importDefault(SttWebsocket_1).default; } });
var StreamingSTTClient_1 = require("./wrapper/StreamingSTTClient");
Object.defineProperty(exports, "StreamingSTTClient", { enumerable: true, get: function () { return StreamingSTTClient_1.StreamingSTTClient; } });
var environments_1 = require("./environments");
Object.defineProperty(exports, "CartesiaEnvironment", { enumerable: true, get: function () { return environments_1.CartesiaEnvironment; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "CartesiaError", { enumerable: true, get: function () { return errors_1.CartesiaError; } });
Object.defineProperty(exports, "CartesiaTimeoutError", { enumerable: true, get: function () { return errors_1.CartesiaTimeoutError; } });
