// @cartesia/cartesia-js/serialization/resources/tts/types/TtsRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TtsRequest = void 0;
const core = __importStar(require("../../../../core"));
const TtsRequestVoiceSpecifier_1 = require("./TtsRequestVoiceSpecifier");
const SupportedLanguage_1 = require("./SupportedLanguage");
const OutputFormat_1 = require("./OutputFormat");
const GenerationConfig_1 = require("./GenerationConfig");
const ModelSpeed_1 = require("./ModelSpeed");
exports.TtsRequest = core.serialization.object({
    modelId: core.serialization.property("model_id", core.serialization.string()),
    transcript: core.serialization.string(),
    voice: TtsRequestVoiceSpecifier_1.TtsRequestVoiceSpecifier,
    language: SupportedLanguage_1.SupportedLanguage.optional(),
    outputFormat: core.serialization.property("output_format", OutputFormat_1.OutputFormat),
    generationConfig: core.serialization.property("generation_config", GenerationConfig_1.GenerationConfig.optional()),
    duration: core.serialization.number().optional(),
    speed: ModelSpeed_1.ModelSpeed.optional(),
    pronunciationDictId: core.serialization.property("pronunciation_dict_id", core.serialization.string().optional()),
});
