// @cartesia/cartesia-js/serialization/resources/tts/types/WebSocketTtsOutput.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketTtsOutput = void 0;
const core = __importStar(require("../../../../core"));
const WordTimestamps_1 = require("./WordTimestamps");
const PhonemeTimestamps_1 = require("./PhonemeTimestamps");
const ContextId_1 = require("./ContextId");
const FlushId_1 = require("./FlushId");
exports.WebSocketTtsOutput = core.serialization.object({
    wordTimestamps: core.serialization.property("word_timestamps", WordTimestamps_1.WordTimestamps.optional()),
    phonemeTimestamps: core.serialization.property("phoneme_timestamps", PhonemeTimestamps_1.PhonemeTimestamps.optional()),
    audio: core.serialization.unknown().optional(),
    contextId: core.serialization.property("context_id", ContextId_1.ContextId.optional()),
    flushId: core.serialization.property("flush_id", FlushId_1.FlushId.optional()),
    flushDone: core.serialization.property("flush_done", core.serialization.boolean().optional()),
});
