// @cartesia/cartesia-js/wrapper/SttWebsocket.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SttWebsocket_instances, _SttWebsocket_isConnected, _SttWebsocket_model, _SttWebsocket_language, _SttWebsocket_encoding, _SttWebsocket_sampleRate, _SttWebsocket_minVolume, _SttWebsocket_maxSilenceDurationSecs, _SttWebsocket_connectionPromise, _SttWebsocket_ensureConnected;
Object.defineProperty(exports, "__esModule", { value: true });
const core = __importStar(require("../core"));
const environments = __importStar(require("../environments"));
const websocket_1 = require("../core/websocket");
class SttWebsocket {
    constructor({ model = "ink-whisper", language = "en", encoding, sampleRate, minVolume, maxSilenceDurationSecs, }, options) {
        _SttWebsocket_instances.add(this);
        this.options = options;
        _SttWebsocket_isConnected.set(this, false);
        _SttWebsocket_model.set(this, void 0);
        _SttWebsocket_language.set(this, void 0);
        _SttWebsocket_encoding.set(this, void 0);
        _SttWebsocket_sampleRate.set(this, void 0);
        _SttWebsocket_minVolume.set(this, void 0);
        _SttWebsocket_maxSilenceDurationSecs.set(this, void 0);
        _SttWebsocket_connectionPromise.set(this, void 0);
        if (!model) {
            throw new Error("model parameter is required");
        }
        if (!encoding) {
            throw new Error("encoding parameter is required");
        }
        if (!sampleRate) {
            throw new Error("sampleRate parameter is required");
        }
        __classPrivateFieldSet(this, _SttWebsocket_model, model, "f");
        __classPrivateFieldSet(this, _SttWebsocket_language, language, "f");
        __classPrivateFieldSet(this, _SttWebsocket_encoding, encoding, "f");
        __classPrivateFieldSet(this, _SttWebsocket_sampleRate, sampleRate, "f");
        __classPrivateFieldSet(this, _SttWebsocket_minVolume, minVolume, "f");
        __classPrivateFieldSet(this, _SttWebsocket_maxSilenceDurationSecs, maxSilenceDurationSecs, "f");
    }
    send(audioData) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            yield __classPrivateFieldGet(this, _SttWebsocket_instances, "m", _SttWebsocket_ensureConnected).call(this);
            (_a = this.socket) === null || _a === void 0 ? void 0 : _a.send(audioData);
        });
    }
    finalize() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            yield __classPrivateFieldGet(this, _SttWebsocket_instances, "m", _SttWebsocket_ensureConnected).call(this);
            (_a = this.socket) === null || _a === void 0 ? void 0 : _a.send("finalize");
        });
    }
    done() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            yield __classPrivateFieldGet(this, _SttWebsocket_instances, "m", _SttWebsocket_ensureConnected).call(this);
            (_a = this.socket) === null || _a === void 0 ? void 0 : _a.send("done");
        });
    }
    onMessage(callback) {
        return __awaiter(this, void 0, void 0, function* () {
            yield __classPrivateFieldGet(this, _SttWebsocket_instances, "m", _SttWebsocket_ensureConnected).call(this);
            if (!this.socket) {
                throw new Error("WebSocket failed to initialize after connection attempt.");
            }
            this.socket.addEventListener("message", (event) => {
                try {
                    const data = JSON.parse(event.data);
                    const result = {
                        type: data.type,
                        requestId: data.request_id || "",
                    };
                    // Handle different message types according to stt.yml spec
                    if (data.type === "transcript") {
                        result.text = data.text || "";
                        result.isFinal = data.is_final || false;
                        result.duration = data.duration;
                        result.language = data.language;
                        // Include word-level timestamps if available
                        if (data.words) {
                            result.words = data.words;
                        }
                    }
                    else if (data.type === "flush_done") {
                        // Acknowledgment for finalize command
                        // Only requestId is needed for flush_done
                    }
                    else if (data.type === "done") {
                        // Acknowledgment for done command - session complete
                        // Only requestId is needed for done
                    }
                    else if (data.type === "error") {
                        result.message = data.message;
                    }
                    callback(result);
                }
                catch (error) {
                    callback({
                        type: "error",
                        requestId: "",
                        message: `Failed to parse message: ${error}`,
                    });
                }
            });
        });
    }
    connect() {
        return __awaiter(this, arguments, void 0, function* (connectOptions = {}) {
            if (__classPrivateFieldGet(this, _SttWebsocket_isConnected, "f")) {
                throw new Error("WebSocket is already connected.");
            }
            this.socket = new websocket_1.ReconnectingWebSocket(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const baseUrl = ((_a = (yield core.Supplier.get(this.options.environment))) !== null && _a !== void 0 ? _a : environments.CartesiaEnvironment.Production).replace(/^https?:\/\//, "");
                const params = {
                    model: __classPrivateFieldGet(this, _SttWebsocket_model, "f"),
                    cartesia_version: this.options.cartesiaVersion || "2024-06-10",
                    encoding: __classPrivateFieldGet(this, _SttWebsocket_encoding, "f"),
                    sample_rate: __classPrivateFieldGet(this, _SttWebsocket_sampleRate, "f").toString(),
                };
                if (__classPrivateFieldGet(this, _SttWebsocket_language, "f"))
                    params.language = __classPrivateFieldGet(this, _SttWebsocket_language, "f");
                if (__classPrivateFieldGet(this, _SttWebsocket_minVolume, "f") !== undefined)
                    params.min_volume = __classPrivateFieldGet(this, _SttWebsocket_minVolume, "f").toString();
                if (__classPrivateFieldGet(this, _SttWebsocket_maxSilenceDurationSecs, "f") !== undefined)
                    params.max_silence_duration_secs = __classPrivateFieldGet(this, _SttWebsocket_maxSilenceDurationSecs, "f").toString();
                const apiKey = yield core.Supplier.get(this.options.apiKey);
                if (apiKey) {
                    params.api_key = apiKey;
                }
                else if (connectOptions.accessToken) {
                    params.access_token = connectOptions.accessToken;
                }
                const queryString = Object.keys(params)
                    .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
                    .join('&');
                return `wss://${baseUrl}/stt/websocket?${queryString}`;
            }), undefined, connectOptions);
            return new Promise((resolve, reject) => {
                this.socket.onopen = () => {
                    __classPrivateFieldSet(this, _SttWebsocket_isConnected, true, "f");
                    resolve();
                };
                this.socket.onclose = () => {
                    __classPrivateFieldSet(this, _SttWebsocket_isConnected, false, "f");
                };
                this.socket.onerror = () => {
                    reject(new Error("WebSocket failed to connect."));
                };
                this.socket.reconnect();
            });
        });
    }
    disconnect() {
        var _a;
        (_a = this.socket) === null || _a === void 0 ? void 0 : _a.close();
        __classPrivateFieldSet(this, _SttWebsocket_isConnected, false, "f");
    }
}
_SttWebsocket_isConnected = new WeakMap(), _SttWebsocket_model = new WeakMap(), _SttWebsocket_language = new WeakMap(), _SttWebsocket_encoding = new WeakMap(), _SttWebsocket_sampleRate = new WeakMap(), _SttWebsocket_minVolume = new WeakMap(), _SttWebsocket_maxSilenceDurationSecs = new WeakMap(), _SttWebsocket_connectionPromise = new WeakMap(), _SttWebsocket_instances = new WeakSet(), _SttWebsocket_ensureConnected = function _SttWebsocket_ensureConnected() {
    return __awaiter(this, void 0, void 0, function* () {
        if (__classPrivateFieldGet(this, _SttWebsocket_isConnected, "f"))
            return;
        if (!__classPrivateFieldGet(this, _SttWebsocket_connectionPromise, "f")) {
            __classPrivateFieldSet(this, _SttWebsocket_connectionPromise, this.connect(), "f");
        }
        yield __classPrivateFieldGet(this, _SttWebsocket_connectionPromise, "f");
    });
};
exports.default = SttWebsocket;
