// @elevenlabs/elevenlabs-js/serialization/types/AgentBranchSummary.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentBranchSummary = void 0;
const core = __importStar(require("../../core"));
const BranchProtectionStatus_1 = require("./BranchProtectionStatus");
const ResourceAccessInfo_1 = require("./ResourceAccessInfo");
exports.AgentBranchSummary = core.serialization.object({
    id: core.serialization.string(),
    name: core.serialization.string(),
    agentId: core.serialization.property("agent_id", core.serialization.string()),
    description: core.serialization.string(),
    createdAt: core.serialization.property("created_at", core.serialization.number()),
    lastCommittedAt: core.serialization.property("last_committed_at", core.serialization.number()),
    isArchived: core.serialization.property("is_archived", core.serialization.boolean()),
    protectionStatus: core.serialization.property("protection_status", BranchProtectionStatus_1.BranchProtectionStatus.optional()),
    accessInfo: core.serialization.property("access_info", ResourceAccessInfo_1.ResourceAccessInfo.optional()),
    currentLivePercentage: core.serialization.property("current_live_percentage", core.serialization.number().optional()),
    draftExists: core.serialization.property("draft_exists", core.serialization.boolean().optional()),
});
