// @elevenlabs/elevenlabs-js/serialization/types/ApiIntegrationWebhookToolConfigOutput.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiIntegrationWebhookToolConfigOutput = void 0;
const core = __importStar(require("../../core"));
const ApiIntegrationWebhookOverridesOutput_1 = require("./ApiIntegrationWebhookOverridesOutput");
const DynamicVariableAssignment_1 = require("./DynamicVariableAssignment");
const DynamicVariablesConfig_1 = require("./DynamicVariablesConfig");
const ToolCallSoundBehavior_1 = require("./ToolCallSoundBehavior");
const ToolCallSoundType_1 = require("./ToolCallSoundType");
const ToolErrorHandlingMode_1 = require("./ToolErrorHandlingMode");
const ToolExecutionMode_1 = require("./ToolExecutionMode");
exports.ApiIntegrationWebhookToolConfigOutput = core.serialization.object({
    name: core.serialization.string(),
    description: core.serialization.string(),
    responseTimeoutSecs: core.serialization.property("response_timeout_secs", core.serialization.number()),
    disableInterruptions: core.serialization.property("disable_interruptions", core.serialization.boolean()),
    forcePreToolSpeech: core.serialization.property("force_pre_tool_speech", core.serialization.boolean()),
    assignments: core.serialization.list(DynamicVariableAssignment_1.DynamicVariableAssignment),
    toolCallSound: core.serialization.property("tool_call_sound", ToolCallSoundType_1.ToolCallSoundType.optional()),
    toolCallSoundBehavior: core.serialization.property("tool_call_sound_behavior", ToolCallSoundBehavior_1.ToolCallSoundBehavior),
    toolErrorHandlingMode: core.serialization.property("tool_error_handling_mode", ToolErrorHandlingMode_1.ToolErrorHandlingMode),
    dynamicVariables: core.serialization.property("dynamic_variables", DynamicVariablesConfig_1.DynamicVariablesConfig),
    executionMode: core.serialization.property("execution_mode", ToolExecutionMode_1.ToolExecutionMode),
    toolVersion: core.serialization.property("tool_version", core.serialization.string()),
    apiIntegrationId: core.serialization.property("api_integration_id", core.serialization.string()),
    apiIntegrationConnectionId: core.serialization.property("api_integration_connection_id", core.serialization.string()),
    apiSchemaOverrides: core.serialization.property("api_schema_overrides", ApiIntegrationWebhookOverridesOutput_1.ApiIntegrationWebhookOverridesOutput.optional()),
});
