// @elevenlabs/elevenlabs-js/serialization/types/ProjectExternalAudioResponseModel.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectExternalAudioResponseModel = void 0;
const core = __importStar(require("../../core"));
const PendingClipTask_1 = require("./PendingClipTask");
exports.ProjectExternalAudioResponseModel = core.serialization.object({
    externalAudioId: core.serialization.property("external_audio_id", core.serialization.string()),
    filename: core.serialization.string(),
    signedUrl: core.serialization.property("signed_url", core.serialization.string()),
    offsetMs: core.serialization.property("offset_ms", core.serialization.number()),
    durationMs: core.serialization.property("duration_ms", core.serialization.number()),
    startTimeMs: core.serialization.property("start_time_ms", core.serialization.number()),
    endTimeMs: core.serialization.property("end_time_ms", core.serialization.number()),
    order: core.serialization.string(),
    trackId: core.serialization.property("track_id", core.serialization.string()),
    createdAtMs: core.serialization.property("created_at_ms", core.serialization.number()),
    updatedAtMs: core.serialization.property("updated_at_ms", core.serialization.number()),
    volumeGainDb: core.serialization.property("volume_gain_db", core.serialization.number().optional()),
    muted: core.serialization.boolean().optional(),
    fadeInMs: core.serialization.property("fade_in_ms", core.serialization.number().optional()),
    fadeOutMs: core.serialization.property("fade_out_ms", core.serialization.number().optional()),
    sourceExternalAudioId: core.serialization.property("source_external_audio_id", core.serialization.string().optional()),
    sourceAssetId: core.serialization.property("source_asset_id", core.serialization.string().optional()),
    pendingBlockIds: core.serialization.property("pending_block_ids", core.serialization.list(core.serialization.string())),
    importSpeechProgress: core.serialization.property("import_speech_progress", core.serialization.number().optional()),
    speechImported: core.serialization.property("speech_imported", core.serialization.boolean().optional()),
    dubAudioProgress: core.serialization.property("dub_audio_progress", core.serialization.number().optional()),
    pendingTask: core.serialization.property("pending_task", PendingClipTask_1.PendingClipTask.optional()),
    currentSnapshotId: core.serialization.property("current_snapshot_id", core.serialization.string().optional()),
});
