// @elevenlabs/elevenlabs-js/serialization/types/SessionStartedPayloadConfig.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionStartedPayloadConfig = void 0;
const core = __importStar(require("../../core"));
const AudioFormatEnum_1 = require("./AudioFormatEnum");
const SessionStartedPayloadConfigCommitStrategy_1 = require("./SessionStartedPayloadConfigCommitStrategy");
exports.SessionStartedPayloadConfig = core.serialization.object({
    sampleRate: core.serialization.property("sample_rate", core.serialization.number().optional()),
    audioFormat: core.serialization.property("audio_format", AudioFormatEnum_1.AudioFormatEnum.optional()),
    languageCode: core.serialization.property("language_code", core.serialization.string().optional()),
    commitStrategy: core.serialization.property("commit_strategy", SessionStartedPayloadConfigCommitStrategy_1.SessionStartedPayloadConfigCommitStrategy.optional()),
    vadSilenceThresholdSecs: core.serialization.property("vad_silence_threshold_secs", core.serialization.number().optional()),
    vadThreshold: core.serialization.property("vad_threshold", core.serialization.number().optional()),
    minSpeechDurationMs: core.serialization.property("min_speech_duration_ms", core.serialization.number().optional()),
    minSilenceDurationMs: core.serialization.property("min_silence_duration_ms", core.serialization.number().optional()),
    modelId: core.serialization.property("model_id", core.serialization.string().optional()),
    enableLogging: core.serialization.property("enable_logging", core.serialization.boolean().optional()),
    includeTimestamps: core.serialization.property("include_timestamps", core.serialization.boolean().optional()),
    includeLanguageDetection: core.serialization.property("include_language_detection", core.serialization.boolean().optional()),
});
